ex3/m_depontes/main.py

```diff
@@ -0,0 +1,473 @@
+        """
+        self.M = M
+        self.w = None  # 重み
+        self.path = path
+        self.data = np.array([])
+        self.model = model
+
+    def OLS(self):
+        """OLS計算を行う関数.
+
+        最小二乗法を用いて重みを計算する.
+        モデルを指定することで，Ridge回帰，Lasso回帰，Elastic Net回帰を行うことができる.
+
+        入力：
+            data(np.ndarray): 入力値, shape = (入力値,入力値に対する出力値)
+            M(int): 次元数
+
+        出力：
+            w: 重み
+        """
+        x = self.data[0]  # 入力値
+        y = self.data[1]  # 出力値
+
+        # 入力行列 X の構築 (バイアス項を含む)
+        X = np.vstack([x**i for i in range(self.M)]).T  # shape = (サンプル数, M)
+        # 正規方程式を用いて重みを計算
+        match self.model:
+            case "normal":
+                w = np.linalg.inv(X.T @ X) @ X.T @ y
+                return w
+            case "Ridge":
+                alpha = 0.1  # 正則化パラメータ
+                w = np.linalg.inv(X.T @ X + alpha * np.eye(X.shape[1])) @ X.T @ y
+                return w
+            case "Lasso":
+                alpha = 0.1  # 正則化パラメータ
+                w = self.elastic_net_coordinate_descent(X, y, alpha)
+                return w
+            case "Elastic":
+                alpha = 0.1  # 正則化パラメータ
+                l1_ratio = 0.5  # L1正則化の比率（0～1）
+                w = self.elastic_net_coordinate_descent(X, y, alpha, l1_ratio)
+                return w
+
+    def elastic_net_coordinate_descent(
+        self, X, y, alpha, l1_ratio=1.0, max_iter=1000, tol=1e-4
+    ):
+        """Elastic Net回帰及びLasso回帰を座標降下法で解く関数.
+
+        入力：
+            X(np.ndarray): 入力行列, shape = (サンプル数, 特徴量数)
+            y(np.ndarray): 出力値, shape = (サンプル数,)
+            alpha(float): 正則化パラメータ
+            l1_ratio(float): L1正則化の比率（0～1, デフォルトは1.0でLasso回帰）
+            max_iter(int): 最大反復回数
+            tol(float): 収束判定の閾値
+
+        出力：
+            w(np.ndarray): 重みベクトル, shape = (特徴量数,)
+        """
+        m, n = X.shape
+        w = np.zeros(n)  # 初期重みをゼロで初期化
+
+        for iteration in range(max_iter):
+            w_old = w.copy()
+            for j in range(n):
+                # 残差を計算
+                residual = y - (X @ w - X[:, j] * w[j])
+                # 重みの更新
+                rho = X[:, j].T @ residual
+                l1_term = alpha * l1_ratio
+                l2_term = alpha * (1 - l1_ratio)
+                if rho < -l1_term:
+                    w[j] = (rho + l1_term) / (X[:, j].T @ X[:, j] + l2_term)
+                elif rho > l1_term:
+                    w[j] = (rho - l1_term) / (X[:, j].T @ X[:, j] + l2_term)
+                else:
+                    w[j] = 0
+
+            # 収束判定
+            if np.linalg.norm(w - w_old, ord=2) < tol:
+                break
+
+        return w
+
+    def fit(self) -> None:
+        """データを読み込み、線形回帰を行う関数.
+
+        入力：
+            path(str): データファイルのパス
+            M(int): 次数
+
+        出力：
+            w(np.ndarray): 重み, shape = (次数)
+        """
+        self.data = read_csv(self.path)
+
+        self.w = self.OLS(self.M, self.data)
```

Comment:
```suggestion
        self.w = self.OLS()
```
OLSには引数を渡していないのに引数付きで呼んでいます。